src/mrinufft/operators/base.py

```diff
@@ -115,136 +117,284 @@ class or instance of class if args or kwargs are given.
-            output_ = output
-
-        if output_ is not None:
-            if not (
-                (is_host_array(data_) and is_host_array(output_))
-                or (is_cuda_array(data_) and is_cuda_array(output_))
-            ):
-                raise ValueError(
-                    "input data and output should be " "on the same memory space."
-                )
-        ret_ = fun(self, data_, output_, *args, **kwargs)
+    def wrapper(*args, **kwargs):
+        args = _get_args(fun, args, kwargs)
 
-        if xp.__name__ == "torch" and is_cuda_array(data):
-            return xp.as_tensor(ret_, device=data.device)
+        # get array module from non-self argument
+        xp = _get_array_module(args[data_arg_idx])
 
-        if xp.__name__ == "torch":
-            if data.is_cpu:
-                return xp.from_numpy(ret_)
-            return xp.from_numpy(ret_).to(data.device)
+        # convert all to cupy / numpy according to data arg device
+        args = _to_numpy_cupy(args, data_arg_idx)
 
-        return ret_
+        # run function
+        ret_ = fun(*args)
+
+        # convert output to original array module and device
+        return _to_interface(ret_, xp)
 
     return wrapper
 
 
 def with_torch(fun):
     """Ensure the function works internally with Torch."""
+    if _ismethod(fun):
+        data_arg_idx = 1
+    else:
+        data_arg_idx = 0
 
     @wraps(fun)
-    def wrapper(self, data, output=None, *args, **kwargs):
-        xp = get_array_module(data)
-
-        if xp.__name__ == "numpy":
-            data_ = torch.from_numpy(data)
-            output_ = torch.from_numpy(output) if output is not None else None
-        elif xp.__name__ == "cupy":
-            data_ = torch.from_dlpack(data)
-            output_ = torch.from_dlpack(output) if output is not None else None
-        else:
-            data_ = data
-            output_ = output
+    def wrapper(*args, **kwargs):
+        args = _get_args(fun, args, kwargs)
 
-        ret_ = fun(self, data_, output_, *args, **kwargs)
+        # get array module from non-self argument
+        xp = _get_array_module(args[data_arg_idx])
 
-        if xp.__name__ == "cupy":
-            return cp.from_dlpack(ret_)
-        elif xp.__name__ == "numpy":
-            return ret_.to("cpu").numpy()
+        # convert all to tensorflow
+        args = _to_torch(*args)
 
-        return ret_
+        # run function
+        ret_ = fun(*args)
+
+        # convert output to original array module and device
+        return _to_interface(ret_, xp)
 
     return wrapper
 
 
+def _ismethod(fun):  # ismethod works on instance methods, not classes (always False)
+    first_arg = list(inspect.signature(fun).parameters)[0]
+    if first_arg in ["self", "cls"]:
+        return True
+    else:
+        return False
+
+
+def _get_array_module(input):  # handle native Python case
+    try:
+        return get_array_module(input)
+    except Exception:
+        return np
```

Comment:
I think we could live happily with `get_array_module` returning numpy in the default case as well (and forget about the exception altogether)